[[ CodeResources ]] extension code resource loading

montegoulding opened the pull request.

montegoulding committed:
[[ CodeResources ]] Impmement `revLibraryMapping` property

This patch implements a `revLibraryMapping` array property to get and
set the mapping path for a library. This allows the IDE to have full
control over the where to find the libraries which is required because
on some systems libraries must be placed in certain locations rather
than inside the extension package.

montegoulding committed:
[[ CodeResources ]] Allow full paths for library mappings

This patch allows full paths to be used for library mappings by checking
if the mapping is prefixed with `./` we can determine if it is intended
to be a path relative to `MCappcodepath`. If it is not prefixed then
the mapping path is passed directly to `MCS_resolvepath`. Currently
all relative mappings are already prefixed by `./` so this is a
backwards compatible change.

montegoulding committed:
[[ CodeResources ]] Load from library mapping

This patch loads extension code resources from lbrary mappings
if they have been mapped by the IDE or standalone deploy params.
If there is no mapping then it falls back to try loading the library
by name followed by attempting to load the library using legacy
paths.

This patch also ensures that iOS device builds retain the engine
as the library handle rather than trying to load a library.

montegoulding committed:
[[ CodeResources ]] Implement platform_id variable in gyp

This patch implements a `platform_id` variable in gyp that conforms
to the platform ID spec.

montegoulding committed:
[[ CodeResources ]] Add copy param to build-extension-ios

This patch adds an additional option to the `build-extension-ios.sh`
script to copy the built library to the specified path.

montegoulding committed:
[[ CodeResources ]] Implement example extension with code resource

This patch implements a library extension that includes a code resource.
The extension is an INI to LiveCode array parser using the inih library.

As extensions are not built when building mobile platforms this patch
adds a target for cross platform extension dependencies that will be
built for all platforms.

montegoulding committed:
[[ CodeResources ]] Move callback setter

This patch moves the call to `MCScriptSetLoadLibraryCallback` to ensure it is
called no matter how the extension is loaded. It is also only necessary to call
it once in `MCExtensionInitialize` instead of every time an extension is loaded.

montegoulding committed:
[[ CodeResources ]] Support module code resources in standalone builder

This patch adds support for module code resources to the standalone builder
by copying the resources (or in the case of iOS static linking them when
required) and setting the approriate library mapping so that they can be
found when foreign handlers are being bound.

montegoulding committed:
[[ CodeResources ]] Add release note

montegoulding committed:
[[ CodeResources ]] Add test

montegoulding committed:
[[ CodeResources ]] Try relative paths

This patch has a second attempt at loading an extenral or support library
by appending `./` to indicate it is a relative path and the app code path
should be appended.

montegoulding committed:
[[ CodeResources ]] Add standalone test

montegoulding committed:
[[ CodeResources ]] Convert i386 and x64 to correct arch string

montegoulding committed:
[[ Bug ]] lowercase commercial in standalone name

montegoulding commented:
The test PR that passes

livecodeali commented on engine/src/exec-extension.cpp:
```diff
@@ -229,64 +229,45 @@ void MCEngineLoadExtensionFromData(MCExecContext& ctxt, MCDataRef p_extension_da
+    
+    // extension libraries should be mapped by the IDE or deploy params
+    MCAutoStringRef t_mapped_path;
+    if (MCdispatcher->fetchlibrarymapping(p_name, &t_mapped_path))
```
I presume the `t_mapped_path` is unused because you are just checking if there exists a library mapping here? If so could you add a comment to that effect (or an `MCDispatch::haslibrarymapping`) so we know it's intended

montegoulding replied:
good idea

livecodeali commented:
@montegoulding Other than that minor comment, this looks great. Will mark reviewed once you've fixed the conflict.

montegoulding commented:
OK, will get that done in a sec

montegoulding committed:
[[ CodeResources ]] Implement haslibrarymapping

This patch implements `MCDispatch::haslibrarymapping` to improve code
clarity when we only need to know if a mapping exists rather than its
value.

livecodeali commented:
@livecode-vulcan review ok 8954c62

livecode-vulcan commented:
💙 review by @livecodeali ok 8954c62c3df63cb1a888d5a09f87ae13d4e2f7f4

livecode-vulcan commented:
😞 test failure 8954c62c3df63cb1a888d5a09f87ae13d4e2f7f4

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1028)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1044)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1018)
* try-community-universal-ios-iphoneos11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/91)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/91)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1023)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1045)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1028)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1044)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1027)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1037)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1060)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1023)

livecodeali commented:
@livecode-vulcan retry 8954c62

livecode-vulcan commented:
😞 test failure 8954c62c3df63cb1a888d5a09f87ae13d4e2f7f4

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1038)
* try-community-armv6-android-api9: [pending](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1054)
* try-community-js-emscripten-sdk1.35: [pending](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1028)
* try-community-universal-ios-iphoneos11.1: [pending](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/101)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/101)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1033)
* try-community-universal-mac-macosx10.9: [pending](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1055)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1038)
* try-community-x86-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1054)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1037)
* try-community-x86_64-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1047)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1070)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1033)

montegoulding commented:
There seems to be an odd issue with variable evaluation.

montegoulding committed:
[[ CodeResources ]] Remove options section from gyp platform_id var

This path removes the options section from the platform_id variable on
windows. There is an issue that the configurations are resolved last and
therefore the CRT mode can not be determined before the variable needs
to be used. A TODO note as been included to add the options section when
possible.

livecodeali commented:
@livecode-vulcan review ok 3aefffa

livecode-vulcan commented:
💙 review by @livecodeali ok 3aefffaf3981608593085cd30bc526040f79d738

livecode-vulcan commented:
😎 test success 3aefffaf3981608593085cd30bc526040f79d738

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1080)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1096)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1070)
* try-community-universal-ios-iphoneos11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.2/builds/23)
* try-community-universal-ios-iphonesimulator11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.2/builds/23)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1075)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1097)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1080)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1096)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1079)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1089)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1112)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1075)
